pkg/virt-handler/device-manager/common.go

```
@@ -0,0 +1,123 @@
+)
+
+type DeviceHandler interface {
+	GetDeviceIOMMUGroup(basepath string, pciAddress string) (string, error)
+	GetDeviceDriver(basepath string, pciAddress string) (string, error)
+	GetDeviceNumaNode(basepath string, pciAddress string) (numaNode int)
+	GetDevicePCIID(basepath string, pciAddress string) (string, error)
+	GetMdevParentPCIAddr(mdevUUID string) (string, error)
+}
+
+type DeviceUtilsHandler struct{}
+
+var Handler DeviceHandler
+
+// getDeviceIOMMUGroup gets devices iommu_group
+// e.g. /sys/bus/pci/devices/0000\:65\:00.0/iommu_group -> ../../../../../kernel/iommu_groups/45
+func (h *DeviceUtilsHandler) GetDeviceIOMMUGroup(basepath string, pciAddress string) (string, error) {
+	iommuLink := filepath.Join(basepath, pciAddress, "iommu_group")
+	iommuPath, err := os.Readlink(iommuLink)
+	if err != nil {
+		log.DefaultLogger().Reason(err).Errorf("failed to read iommu_group link %s for device %s", iommuLink, pciAddress)
+		return "", err
+	}
+	_, iommuGroup := filepath.Split(iommuPath)
+	return iommuGroup, nil
+}
+
+// gets device driver
+func (h *DeviceUtilsHandler) GetDeviceDriver(basepath string, pciAddress string) (string, error) {
+	driverLink := filepath.Join(basepath, pciAddress, "driver")
+	driverPath, err := os.Readlink(driverLink)
+	if err != nil {
+		log.DefaultLogger().Reason(err).Errorf("failed to read driver link %s for device %s", driverLink, pciAddress)
+		return "", err
+	}
+	_, driver := filepath.Split(driverPath)
+	return driver, nil
+}
+
+func (h *DeviceUtilsHandler) GetDeviceNumaNode(basepath string, pciAddress string) (numaNode int) {
+	numaNode = -1
+	numaNodePath := filepath.Join(basepath, pciAddress, "numa_node")
+	numaNodeStr, err := ioutil.ReadFile(numaNodePath)
+	if err != nil {
+		log.DefaultLogger().Reason(err).Errorf("failed to read numa_node %s for device %s", numaNodePath, pciAddress)
+		return
+	}
+	numaNodeStr = bytes.TrimSpace(numaNodeStr)
+	numaNode, err = strconv.Atoi(string(numaNodeStr))
+	if err != nil {
+		return
+	}
```

just want to double check it's safe to ignore the errors here.  Is the device simply not advertised if the numaNode isn't detected? so numaNode <= 0

---

yes. if we didn't find a numa node for whatever reason here, we shuoldn't then set a topology hint for the k8s topology manager.
I think -1 is fine.

---

makes sense.

maybe we should throw a log message in for that second error case in the event that somehow the numa node id we parse doesn't happen to be a string?

---

Yea, I'll add now.